packages/sitecore-jss/src/personalization/layout-personalization-context.ts

```diff
@@ -0,0 +1,70 @@
+import { ComponentRendering } from '../layout/models';
+
+export interface SitecorePersonalizationContextState {
+  /**
+   * A value that indicates whether a route is tracked
+   */
+  isTracked: boolean;
+
+  getPersonalizedComponent(componentUid: string): ComponentRendering | null | undefined;
+
+  isLoading(componentUid: string): boolean;
+
+  ensurePersonalizedComponentLoaded(componentUid: string): Promise<ComponentRendering | null>;
+}
+
+export class SitecorePersonalizationContext implements SitecorePersonalizationContextState {
+  public isTracked: boolean;
+  private components?: { [key: string]: ComponentRendering | null };
+  private personalizationOperation: Promise<{
+    [key: string]: ComponentRendering | null;
+  }>;
+
+  constructor(
+    personalizationOperation: Promise<{
+      [key: string]: ComponentRendering | null;
+    }>,
+    isTracked: boolean
+  ) {
+    this.personalizationOperation = personalizationOperation;
+    this.isTracked = isTracked;
+    personalizationOperation.then((components) => {
```

Comment:
Something is wrong here, It's not correct approach to use promises/side-effects in the `constructor`. Think about it.

Reply:
Why this instance can't be created after `personalizationOperation` will be finished?

Reply:
This is by design. 
We start the asynchronous operation of personalization and wrap it with the context to provide an interface for consuming personalization results. In the constructor of the context, we subscribe to the promise resolution to provide later access to its results.